### DISCUSSION TITLE
trouble with morphing vertex in shader

### DISCUSSION THREAD

**Discussion by QUAN-Zhao** [google/filament#5000] [Q&A] [answered]
I use assimp to load my FBX model, and render it out using filament. Now I need to do blendshapes in GPU, I construct a texture for each sub-mesh to store difference of vertices: row represent all vertices for one blendshape, column represent all blendshapes for one vertex. In vertex shader, I read the difference from texture and modified position of vertex, but the render result failed to do the blendshape:

    void materialVertex(inout MaterialVertexInputs material) {
          ...(variable bsId,  bs reference as BlendShape)
          material.worldPosition.rgb += materialParams.bsWeights[bsId] * texture(materialParams_bsMap, vec2(float(getVertexIndex()) / float(materialParams.bsMapWidth), float(bsId) / float(materialParams.bsMapHeight)));
    }

Do I need do some work for world-coordinate-transform?

**Comment by bejado** [comment, marked as answer]
If you're setting `worldPosition` in the vertex shader, then you need to offset it by `getWorldOffset()`. See [heightfield.mat](https://github.com/google/filament/blob/main/samples/materials/heightfield.mat#L22) for an example. Also see our [material documentation](https://google.github.io/filament/Materials.html) (search for `getWorldOffset`).

**Reply by QUAN-Zhao** [reply]
Sorry for the late reply. I tried many composition of getWorldOffset(), getWorldFromModelMatrix?, ..., but the result is still wrong. I wonder if I construct a wrong texture(the texture to save differences of vertices in the mesh) or use a wrong way to morph the vertex's position in vertex shader? I construct the texture using Texture::InternalFormat::RGB32F, so will filament clip the differences of vertices into range [0, 1.0]? Do I need to multiply the difference with getWorldFromModelMatrix()? Why should I add the getWorldOffset() to material.worldPosition.xyz?

I tried many composition of these APIs, but get no success... And I wonder, if there exist a way to morph mesh (blendshapes) in the vertex shader provided by filament.

**Reply by bejado** [reply]
No, Filament shouldn't be clamping the texture samples.

You don't need to apply `getWorldOffset()` if you're just offsetting vertices, for example:

```
material.worldPosition.xyz += offset;
```

However, if you want to set an absolute position in the API-level coordinate system, you need to apply `getWorldOffset()`. You can think of it as two separate coordinate systems. Filament's vertex shaders use a "shifted world space" coordinate system, while Filament's C++ API uses "API world space". This might help clarify:

```
vec3 shiftedToApiWorldSpace(const vec3 v) {
    return v + getWorldOffset();
}

vec3 apiToShiftedWorldSpace(const vec3 v) {
    return v - getWorldOffset();
}

void materialVertex(inout MaterialVertexInputs material) {
    float height = texture(materialParams_height, material.uv0).r;

    vec3 shifted = material.worldPosition.xyz;
    vec3 apiPosition = shiftedToApiWorldSpace(shifted);
    apiPosition.y = height;     // height is an API-level coordinate
    material.worldPosition.xyz = apiToShiftedWorldSpace(apiPosition);
}
```

Note that we do provide a morphing API on `RenderableManager`: https://github.com/google/filament/blob/main/filament/include/filament/RenderableManager.h#L456-L464
There's also a PR in review that adds support for more blendshapes.

**Reply by QUAN-Zhao** [reply]
Thanks for your reply, I have finally made the blendshapes.
